app/contracts/projects/unarchive_contract.rb

```diff
@@ -27,16 +27,17 @@
 #++
 
 module Projects
-  class UnarchiveContract < ModelContract
+  class UnarchiveContract < ::BaseContract
```

Comment:
Because it was not used: the model checks were skipped because of `validate_model?` being overridden to return `false`. That leave the `valid?()` method overriden in `ModelContract` only calling `readonly_attributes_unchanged`. And in the `Projects::ArchiveController`, the project instance being passed never has any modifications, so attributes were always left unchanged. So this is not necessary here.

When I tried to reuse the `Archive/Unarchive`-`Contract` in `Projects::UpdateContract`, at first it was not working because of the `readonly_attributes_unchanged` would block modifying anything if the project is archived (because all fields become readonly when the project is archived). I was looking to override `#readonly_attributes_unchanged` to do nothing, and then I realized that I was not using anything from `ModelContract`, so I changed to the parent class of `ModelContract`, which is `::BaseContract`

And then in the `Projects::UpdateService`, it calls `Projects::ArchiveService` or `Projects::UnarchiveService` if the `active` field is being changed. In that case, the contract checks of `Projects::ArchiveContract` or `Projects::UnarchiveContract` have already been run by `Projects::UpdateService`, so no need to use a contract once again. That's why [an EmptyContract is being used](https://github.com/opf/openproject/pull/11853/files#diff-da3e8ab4aa6c0709ac38d2a184dc26e40d7471a15726fd161ddd0c22ed17e86dR102-R105).

It's all in this commit: [`2d2377f`](https://github.com/opf/openproject/pull/11853/commits/2d2377f99dfce64169633d9455e6708d7ead3592)

Also the fact that `allows_to?` was returning false, or that `Users::ProjectRoleCache` was returning no roles when project has `active == false` was annoying to be able to reuse the`Archive/Unarchive`-`Contract` in UpdateContract, because the instance being updated through the API could have the active flag being changed. So I introduced `Project#being_archived?` which checks if the project active state is being changed, so that things still work when this field is being changed. This allowed to remove the [`with_active_assumed` hacky method](https://github.com/opf/openproject/pull/11853/files#diff-969458ad4f527537566e3c9b1ccb9410dd0424d78637f65a8185bbb7d7977fb2L122-L129).

Reply:
Admittedly you are right about this part. As said when we were talking, I misunderstood `::BaseContract` to mean `Projects::BaseContract` which is incorrect.

I have now looked through how the contract is currently used and your assessment seems to be right. But the way the Archive/Unarchive contracts are used currently is different than how we use them in other places where we typically do something like this:

1) change the attributes in a service
2) validate the changes via a contract
3) persist the changes triggered by the service (`model.save`) in case of confirmed validity

Here, the contracts are used only to check for permissions and for checking the current state of the projects and its descendants with the `active` attribute being unchanged. 

That definitely has the benefit of allowing the Contract to be callable from another one (e.g. `UpdateContract` in this case) without the hassle this would cause were the write protection in place.  

If the contract is misused, it will not offer any protection that might be expected by it, e.g.
```
project = Project.find('abc')
project.active = false
project.name = 'xyz'

Projects::ArchiveContract
  .new(project, User.admin.first)
  .valid?
```

would return `true` even though more is changed than what is necessary when archiving because the write protection does not exist. Admittedly, the write protection was too aggressive before because it also did not allow the `active` property to be written.

When arguing with the principle of least surprise in mind, I'd have preferred the contracts to keep inheriting from the `ModelContract`, adding `attribute :active` to the contracts and have the property changed in the service before the contract validates. But that would clash seriously with the way the contracts are now used by the `UpdateContract` which has merits of its own.

TL;DR In conclusion I agree that the approach you took makes sense if only because of the way the contracts are used in this specific case.

Reply:
I understand your concerns.

On the other hand, as you said, the previous version of ArchiveContract was too agressive:
on `dev`

```ruby
project = Project.last
project.active = false
Projects::ArchiveContract.new(project, User.admin.first).valid?
=> false
```

It looks more like a `ArchivableContract`.

Is there some kind of guards we could use to make sure it is not used in an unintended way?

Reply:
An alternative would be to just delete `Projects::ArchiveContract` and `Projects::UnarchiveContract`, and put all the logic in `Projects::UpdateContract`, and this contract would be used for all of `Projects::ArchiveService`, `Projects::UnarchiveService`, and `Projects::UpdateService`.

On the other hand, `Projects::ArchiveService`, `Projects::UnarchiveService` are overriding `#persist` to only update the `active` field.

Reply:
> Is there some kind of guards we could use to make sure it is not used in an unintended way?

Can't think of one currently.